web/content.go

```diff
@@ -0,0 +1,370 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"context"
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"reflect"
+	"strconv"
+)
+
+// acceptTypeAll indicates a accept type from http request.
+// It means client can receive any content.
+// Request content type in header "Content-Type" must not set to "*/*".
+// It only can exist in request header "Accept".
+// In most time, it locate at the last element of "Accept".
+// It's default value if client have not set "Accept" header.
+const acceptTypeAll = "*/*"
+
+// Consumer handles specifically typed data from a reader and unmarshals it into an object.
+type Consumer interface {
+	// ContentType returns a HTTP MIME type.
+	ContentType() string
+	// Consume unmarshals data from r into v.
+	Consume(r io.Reader, v interface{}) error
+}
+
+// Consumer marshals an object to specifically typed data and write it into a writer.
+type Producer interface {
+	// ContentType returns a HTTP MIME type.
+	ContentType() string
+	// Produce marshals v to data and write to w.
+	Produce(w io.Writer, v interface{}) error
+}
+
+var consumers = map[string]Consumer{}
+var producers = map[string]Producer{}
+
+// ConsumerFor gets a consumer for specified content type.
+func ConsumerFor(contentType string) Consumer {
+	return consumers[contentType]
+}
+
+// ProducerFor gets a producer for specified content type.
+func ProducerFor(contentType string) Producer {
+	return producers[contentType]
+}
+
+// RegisterConsumer register a consumer. A consumer must not handle "*/*".
+func RegisterConsumer(c Consumer) error {
+	if c.ContentType() == acceptTypeAll {
+		return fmt.Errorf("must not register a consumer for %s", acceptTypeAll)
+	}
+	if _, ok := consumers[c.ContentType()]; ok {
+		return fmt.Errorf("consumer %s has been registered", c.ContentType())
+	}
+	consumers[c.ContentType()] = c
+	return nil
+}
+
+// RegisterProducer register a producer. A producer must not handle "*/*".
+func RegisterProducer(p Producer) error {
+	if p.ContentType() == acceptTypeAll {
+		return fmt.Errorf("must not register a producer for %s", acceptTypeAll)
+	}
+	if _, ok := producers[p.ContentType()]; ok {
+		return fmt.Errorf("producer %s has been registered", p.ContentType())
+	}
+	producers[p.ContentType()] = p
+	return nil
+}
+
+// JSONSerializer implements Consumer and Producer for content type "application/json".
+type JSONSerializer struct{}
+
+// ContentType returns json MIME type.
+func (s *JSONSerializer) ContentType() string {
+	return "application/json"
+}
+
+// Consume unmarshals json from r into v.
+func (s *JSONSerializer) Consume(r io.Reader, v interface{}) error {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(data, v)
+}
+
+// Produce marshals v to json and write to w.
+func (s *JSONSerializer) Produce(w io.Writer, v interface{}) error {
+	data, err := json.Marshal(v)
```

Comment:
Why not just use `json.Encoder`?

Reply:
The world does not only have json. There are xml, files, and many strange types.